Fix UserManager after EFcore refactor

JPVenson opened the pull request:
Changes:
- Removes cache from UserManager
- Fixes RowVersion mismatch lookup
- Adds locking for UserManager operations based on UserId to avoid cross-operational access

fixes https://github.com/jellyfin/jellyfin/issues/15153
fixes https://github.com/jellyfin/jellyfin/issues/16353
fixes #15714

JPVenson committed:
#15153 fixed

nagisa commented:
> Call for testing is pointless as next to impossible to reproduce reliably as it currently stands.

I get `DbUpdateConcurrencyException`s quite consistently with `supersonic` restarts. I'll apply this patch to my jellyfin sometime later.

JPVenson committed:
Merge remote-tracking branch 'jellyfinorigin/release-10.11.z' into bugfix/15153_FixUserConcurrency

JPVenson committed:
Refactored all UserManager db access methods
Fixed stale cached entities in UserManager
Fixed wrong state persisting though method calls

IDisposable commented on Jellyfin.Server.Implementations/Users/UserManager.cs:
```diff
@@ -139,42 +160,62 @@ public UserManager(
+#pragma warning disable CA1311 // Specify a culture or use an invariant version to avoid implicit dependency on current culture
+#pragma warning disable CA1304 // The behavior of 'string.ToUpper()' could vary based on the current user's locale settings
+            return UserQuery(dbContext)
+                .FirstOrDefault(u => u.Username.ToUpper() == name.ToUpper());
```
Wouldn't it be better to use [`.ToUpperInvariant()`](https://learn.microsoft.com/en-us/dotnet/api/system.string.toupperinvariant?view=net-10.0) here?

JPVenson replied:
i need to check if that is available in an EFQuery but i dont think so

IDisposable replied:
It doesn't, sorry... because the column's collation determines if that matters. Did we create the username with a case-insensitive collation... in which case the database will do the right thing on a simple `.Equals()`... for EFCore, I think best-practice is to use `u.Username.Equals(name, StringComparison.OrdinalIgnoreCase)` because EFCore can recognize it.

JPVenson replied:
Are you sure?

> In addition, .NET provides overloads of [string.Equals](https://learn.microsoft.com/en-us/dotnet/api/system.string.equals) accepting a [StringComparison](https://learn.microsoft.com/en-us/dotnet/api/system.stringcomparison) enum, which allows specifying case-sensitivity and a culture for the comparison. By design, EF Core refrains from translating these overloads to SQL, and attempting to use them will result in an exception.

https://learn.microsoft.com/en-us/ef/core/miscellaneous/collations-and-case-sensitivity

IDisposable replied:
No, I'm not sure. We used to have the opposite guidance... looks like they came to the same realization as I did... the database's collation on the column(s) and index really should drive this.

> To force a query to use case-sensitive or case-insensitive comparison, specify a collation explicitly via EF.Functions.Collate as [detailed above](https://learn.microsoft.com/en-us/ef/core/miscellaneous/collations-and-case-sensitivity#explicit-collations-and-indexes).

Looking briefly at the EFCore code, it looks like it does support some case-insensitivity with EFCore.Cosmos CONTAINS, but not for Equals.

JPVenson replied:
Yea lets not deal with that in this PR. Other parts of the code do that the same and that is a baseline discussion for another time.

IDisposable commented on Jellyfin.Server.Implementations/Users/UserManager.cs:
```diff
@@ -139,42 +160,62 @@ public UserManager(
             ThrowIfInvalidUsername(newName);
 
-            if (user.Username.Equals(newName, StringComparison.Ordinal))
+            if (oldName.Equals(newName, StringComparison.Ordinal))
```
Since we're using `.ToUpper()` above, shouldn't this be `StringComparison.OrdinalIgnoreCase`?

IDisposable commented on Jellyfin.Server.Implementations/Users/UserManager.cs:
```diff
@@ -139,42 +160,62 @@ public UserManager(
-                user.Username = newName;
-                await UpdateUserInternalAsync(dbContext, user).ConfigureAwait(false);
+                    // Load a fresh tracked entity inside the lock so the RowVersion is current.
+                    user = await dbContext.Users
```
Could use `UserQuery(dbContext)` here, except it forces `.AsNoTracking()` might be cleaner to make that an optional flag so we only have the `.Include` train in one place.

IDisposable commented on Jellyfin.Server.Implementations/Users/UserManager.cs:
```diff
@@ -218,23 +258,30 @@ public async Task<User> CreateUserAsync(string name)
+                        .AnyAsync(u => u.Username.ToUpper() == name.ToUpper())
+                        .ConfigureAwait(false))
+                {
+                    throw new ArgumentException(string.Format(
```
Since we're not actually transactional... this condition could still occur unless we have a unique constraint on the username column, in which case all this could simply be done by catching the exception (and ensuring/detecting it's a unique-constraint violation)

JPVenson replied:
yes technically correct but that would be using an exception for control flow which is bad design

IDisposable commented on Jellyfin.Server.Implementations/Users/UserManager.cs:
```diff
@@ -245,39 +292,48 @@ public async Task<User> CreateUserAsync(string name)
+                    var userCount = await dbContext.Users.CountAsync().ConfigureAwait(false);
+                    if (userCount == 1)
+                    {
+                        throw new InvalidOperationException(string.Format(
```
Technically, we don't need to check this, because the very next condition will fail if there won't be a remaining administrator... which is really what we're trying to ensure, right?

JPVenson replied:
Yes but that code was there before and i dont really want to start touching that. There _might_ be a reason for it i dont see rn. I remove it if you insist.

IDisposable reviewed (changes requested):
Comments aren't blocking, but should be considered

JPVenson committed:
Apply review comments

cvium commented on Jellyfin.Server.Implementations/Users/UserManager.cs:
```diff
@@ -403,102 +481,114 @@ public UserDto GetUserDto(User user, string? remoteEndPoint = null)
-                if (success
-                    && authenticationProvider is not null
-                    && authenticationProvider is not DefaultAuthenticationProvider)
+                // Reload the user now that we hold the lock so the RowVersion is current.
```
Doesn't it defeat the purpose of the concurrency check?

JPVenson replied:
No not really well at least not in this particular case. 

The issue with re-using the original dto would be here that it has the old auth state but there is no real reason for us to use that. The authentication is valid or not based on the input parameters. 

The concurrency token should never __override__ data with an older state but as we are only want to modify the object inside the lock anyway we requery it.
What we modify here is essentially only the LastActivityDate and InvalidLoginAttemptCount and those are either static values or increments so its even better if we get the most current values here when we enter the lock again so that if the user has been locked out by a previous locking attempt we DO NOT override it with an previous InvalidLoginAttemptCount +1 but always get the most up to date data.

Works as intended

cvium replied:
But the fact it's even needed to refresh it points to an incorrect usage pattern somewhere. You're ignoring any concurrency which defeats the purpose.

It's not the only place where it's done, it's just where my comment landed. I'm talking more general and not this specific case necessarily

JPVenson replied:
i mean i do not?

If you look at the code the base issue is that User does not have a DTO. That is why we cannot "reuse" the DTO as an entity here. Dont think of the object that comes via the parameter as an entity but as what should be the DTO and if it would be an DTO we would need to requery anyway to get tracking.

I dont disagree there is an issue with an incorrect usage but that is because we are not using Dtos here. Also doing it this way is the absolute correct way of doing it even if that would be an DTO instead of the actual entity.

But this elevates to a bigger issue we have with the legacy code overall esp with users. 

Generally without massive 3 layer architecture refactoring this is the easiest solution to fix it and even if we do it, the code inside the lock stays the same.

JPVenson replied:
I rewrote the ChangePassword and ResetPassword methods so this no longer applies either

JPVenson commented:
@IDisposable if your comments aren't blocking dont submit a "request changes" because that is blocking xD

IDisposable reviewed (approved):
Looks good, not sure how much we care about this comment https://github.com/jellyfin/jellyfin/pull/15368/changes/2ac0edc052ed5d22cee3d46fb933bb12a2b36283..586fa01e4685125ef3a9e71f1070a40e61ef7620#diff-cff9cf8fca144299e27f39f7f6625130687e75c435c23aae1fc35126fd175a1fR695

IDisposable commented:
> @IDisposable if your comments aren't blocking dont submit a "request changes" because that is blocking xD

That was a miss-click... mea culpa.

cvium commented on Jellyfin.Server.Implementations/Users/UserManager.cs:
```diff
@@ -139,42 +160,59 @@ public UserManager(
+                        .ConfigureAwait(false)
+                        ?? throw new ResourceNotFoundException(nameof(userId));
+
+                    user.Username = newName;
```
This is the place I was looking for. A fresh entity is loaded but another process could have changed the username before the lock was acquired, but it will be discarded. Same thing happens in changepassword.

JPVenson replied:
No i am sory to dissapoint you, that comment is just wrong. If you look at the code this is the first time we load the entity here.

JPVenson replied:
Removed the comment

JPVenson committed:
Removed obsolete comment

JPVenson committed:
Update Reset method to not rely on externally available entity
